plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol

```diff
@@ -193,25 +197,45 @@ library PaymentProcessInFlightExit {
         }
     }
 
-    function flagInputsAndOutputsSpent(
+    function flagOutputsWhenNoncanonical(
+        PlasmaFramework framework,
+        PaymentExitDataModel.InFlightExit memory exit,
+        address token
+    )
+        private
+    {
+        uint256 piggybackedInputNumOfTheToken;
+        for (uint16 i = 0; i < MAX_INPUT_NUM; i++) {
+            if (exit.inputs[i].token == token && exit.isInputPiggybacked(i)) {
+                piggybackedInputNumOfTheToken++;
+            }
+        }
+
+        bytes32[] memory outputIdsToFlag = new bytes32[](piggybackedInputNumOfTheToken);
+        uint indexForOutputIds = 0;
+        for (uint16 i = 0; i < MAX_INPUT_NUM; i++) {
+            if (exit.inputs[i].token == token && exit.isInputPiggybacked(i)) {
+                outputIdsToFlag[indexForOutputIds] = exit.inputs[i].outputId;
+                indexForOutputIds++;
+            }
+        }
+        framework.batchFlagOutputsSpent(outputIdsToFlag);
+    }
+
+    function flagOutputsWhenCanonical(
         PlasmaFramework framework,
         PaymentExitDataModel.InFlightExit memory exit,
         address token
     )
         private
     {
-        // We flag _all_ inputs regardless of whether it is piggybacked
-        // If exiting from output, all inputs are considered spent and can only exit from output in future.
-        // If exiting from input, to keep things simple, all users must piggyback the input at the same time,
-        // instead of re-starting the IFE and then re-piggyback their input.
         uint256 inputNumOfTheToken;
         for (uint16 i = 0; i < MAX_INPUT_NUM; i++) {
-            if (exit.inputs[i].token == token && exit.inputs[i].amount > 0) {
```

Comment:
does not seem necessary to check whether amount exists or not? We set data for all inputs and outputs in start IFE no matter the amount value.

Reply:
ah....okay. So we use fixed array in IFE exit data so it is actually needed.